### ISSUE TITLE
Recursive Invariance issue

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2115]
**Describe the bug**
I tried to make a type of nested lists of ints. When I try to use that type on a value of list of int it has invariance issues. The bug disappears if I change to nested sequence of ints.

**To Reproduce**
```python
from typing import List, TypeVar, Union

T = TypeVar('T')

RecList = Union[List['RecList[T]'], T]
RecIntList = RecList[int]
x: RecIntList
y: List[int] = []
x = y
```

Error message shown on the last line is,

Expression of type "list[int]" cannot be assigned to declared type "RecIntList"
  Type "list[int]" cannot be assigned to type "RecIntList"
    "list[int]" is incompatible with "int"
      TypeVar "_T@list" is invariant
        Type "int" cannot be assigned to type "List[RecList] | int"
          "int" is incompatible with "List[RecList]"

**Expected behavior**
No error message. I expect RecIntList to mean int | List[int] | List[List[int]] | ...

**VS Code extension or command-line**
pylance v2021.7.6-pre.1.

**Comment by erictraut** [comment]
This is the correct behavior.

`List[int]` isn't type compatible with `RecIntList` because the type parameter for `List` is invariant. To make this work, you need can either change the type annotation for `y` to `RecIntList` or you can switch from `List` to `Sequence`.

**Comment by hmc-cs-mdrissi** [comment]
I'm confused. RecIntList I understand as int | List[int] | List[List[int]] | ... I expect List[int] to match the second case. What am I missing?

If not how do I write the type,  int | List[int] | List[List[int]] | ...

**Comment by erictraut** [comment]
If you expand out the type alias for `RecIntList`, it's equivalent to `List[RecList[int]] | int`.

You could define `RecList` as follows:
```python
RecList = Union[List['RecList[T]'], List[T], T]
```

**Comment by hmc-cs-mdrissi** [comment]
That makes sense. I don't think your solution is enough though. List[int] would work, but wouldn't List[List[int]] fail for the same reason because,

RecIntList = int | List[int] | List[RecList[int]] = int | List[int] | List[int | List[int] | RecList[int]] = ...

Not sure if there is a good way to write int | List[int] | List[List[int]] | ... Original motivation for this issue was working with a type similar to json type and then I just dropped a couple cases.

**Comment by erictraut** [comment]
This works fine if you assign a List[RecIntList] to `RecIntList`. The problem is that you're trying to assign `List[int]` to `RecIntList`, which is a type error. That's equivalent to assigning a `List[int]` to a `List[int | str]`.

So if you think about this a little differently, you should be able to make this work.

**Comment by hmc-cs-mdrissi** [comment]
What type signature would you use then to make this kind of function type check with each of these callers,

```python
def f(x):
  assert isinstance(x, (int, list))
  if isinstance(x, int):
    return x + 1
  else:
    for index, elem in x:
       x[index] = f(elem)

f(10)
f([1,2,3])
f([[1,2],[3,4]])

# z1/z2 here would normally be parameter of some function that calls f or return type of a value earlier defined in that function. 
z1: List[int]
z2: List[List[int]]
f(z1)
f(z2)

def g() -> List[int]
  ...

f(g())
```

The original error was closest to the last example. I called a function that returned a List[int]. I then wanted to pass that value to my f. Changing the type hint of g to Rec anything would be wrong. g does not think about nested lists and always just returns a list of ints.

I don't think this is a pyright issue though and feels more like recursive types + mutable types cooperate poorly together in python. I consider JSON like types pretty common and useful case for them and if you can't type hint f, then more interesting functions on mutable jsons will also have errors.

**Comment by erictraut** [comment]
With the constraints you've placed on the problem, I don't see a good solution. But I don't think those constraints are required. You just need to think about the problem a little differently.
